river/drift/d3.py

```diff
@@ -0,0 +1,189 @@
+    new. A simple classifier is trained to distinguish these sets. It
+    detects a drift with respect to classifier performance (AUC).
+
+    Parameters
+    ----------
+    old_data_window_size
+        The size of the old data window. (int)
+    new_data_percentage
+        Determines the size of the new data window. It is calculated as
+        a fraction of the old data window. The new data window has size
+        new data percentage * old data window size. (float)
+        From 0.0 to 1.0
+    auc_threshold
+        Required AUC score to signal a drift. (float)
+        From 0.5 to 1.0
+    discriminative_classifier
+        Classifier to be used to distinguish old data from the new data.
+        If None, Logistic Regression with default parameters will be used.
+        It is advised to use a simple model as the goal of this classifier
+        is to determine if the old data and the new data are seperable,
+        not to classify them. (sklearn classifier)
+    seed
+        Is used as as random state for numpy
+
+
+    Examples
+    --------
+    >>> from river import synth
+    >>> from river.drift import D3
+
+    >>> d3 = D3(seed=12345)
+
+    >>> # Simulate a data stream
+    >>> data_stream = synth.Hyperplane(seed=42, n_features=10, mag_change=0.5)
+
+    >>> # Update drift detector and verify if change is detected
+    >>> i = 0
+    >>> for x, y in data_stream.take(250):
+    ...     in_drift, in_warning = d3.update(x)
+    ...     if in_drift:
+    ...         print(f"Change detected at index {i}")
+    ...     i += 1
+    Change detected at index 242
+
+    References
+    ----------
+    [^1]: Ömer Gözüaçık, Alican Büyükçakır, Hamed Bonab, Fazli Can: Unsupervised concept drift detection with a discriminative classifier. CIKM 2019: 2365-2368
+
+    """
+
+    _DISC_CLF_TRAINING_DATA_PERCENTAGE = 0.7
+
+    def __init__(
+        self,
+        old_data_window_size=100,
+        new_data_percentage=0.1,
+        auc_threshold=0.7,
+        discriminative_classifier=None,
+        seed=None,
+    ):
+        super().__init__()
+        self.auc_threshold = auc_threshold
+        self.data_sliding_window = None
+        self.discriminative_classifier = discriminative_classifier
+        self.new_data_percentage = new_data_percentage
+        self.feature_names = None
+        self.old_data_window_size = old_data_window_size
+        self.sliding_window_index = 0
+        self.new_data_window_size = int(
+            self.old_data_window_size * self.new_data_percentage
+        )
+        self.full_window_size = self.old_data_window_size + self.new_data_window_size
+        np.random.seed(seed)
+        super().reset()
+
+    def _is_sliding_window_full(self):
+        """Checks if the sliding window is full."""
+        if self.sliding_window_index < self.full_window_size:
+            return False
+        return True
+
+    def _discriminative_drift_detection(self, old_data, new_data):
```

Comment:
As mentioned, we are trying to minimize dependencies. 

the method uses a batch model in a stream setting. This is not wrong. However, the implementation has some limitations:

- The implementation is restricted to `sklearn` models. The `fit` and `predict_proba` are not standard in `river`, so trying to use a `river` model is not possible.
- There is overhead from converting the data to lists/numpy-arrays.

From the documentation:
> It is advised to use a simple model as the goal of this classifier
> is to determine if the old data and the new data are separable,
> not to classify them.

Then there is no reason to not use the incremental implementation of [`LogisticRegression`](https://riverml.xyz/latest/api/linear-model/LogisticRegression)  in `river`

This is useful because:
- The implementation is compatible with the rest of the codebase.
- No need to convert the data.
- `np.concatenate` is slow. A faster alternative is a fixed length [`deque`](https://docs.python.org/3/library/collections.html#collections.deque).
- The testing (`predict_one`) and metric `update` can be performed inside the same loop.
- Numpy methods can be replaced by standard methods.